### ISSUE TITLE
guzzle errors are not handled and dump out to user on plain white screen/web page

### ISSUE THREAD

**Issue by arduent** [Soneso/stellar-php-sdk#29]
I notice that guzzle errors are dumped out/printed and program terminates. So in a web application you get a blank page with a json guzzle error message like this:

```
{
  "type": "https://stellar.org/horizon-errors/bad_request",
  "title": "Bad Request",
  "status": 400,
  "detail": "The request you sent was invalid in some way.",
  "extras": {
    "invalid_field": "account_id",
    "reason": "Account ID must start with `G` and contain 56 alphanum characters"
  }

```
That's sent to the user in the case of an invalid key, IE:

```
//bogus invalid key
$pubkey = 'GC3CT2B55RPLE6JX2U3SOPZFGSE5A3MYFBFYAXSQCJ5BBYJX5HIBTNIX';


try {
	$account = $sdk->requestAccount($pubkey);
} catch (HorizonRequestException $e) {
	notice(  $invalid_key. ' ('.$e->getMessage().')'.EOL);
}

```

So, in Soneso/StellarSDK/StellarSDK.php:
 
```
public function __construct(string $uri)
{
	$this->serverUri = $uri;
	$this->httpClient = new Client([
		'base_uri' => $uri,
        'exceptions' => false,
	]);
}

```
I guess GuzzleHttp changed "exceptions" option to "http_errors" back in 2015. so this needs to be:

```
public function __construct(string $uri)
{
	$this->serverUri = $uri;
	$this->httpClient = new Client([
		'base_uri' => $uri,
        'http_errors' => false,
	]);
}

```

But then "this should not happen" actually happens. :)

in Soneso/StellarSDK/Responses/ResponseHandler.php:

   ```
 public function handleResponse(ResponseInterface $response, string $requestType, Client $httpClient) : Response {

        $content = $response->getBody()->__toString();

        //print($content);

        // not success
        // this should normally not happen since it will be handled by gruzzle (throwing corresponding gruzzle exception)
        if (300 <= $response->getStatusCode()) {
                throw new RuntimeException($content);
        }

```

So it throws a 500 error and you get unhandled runtime exception in error log.

My solution was to add the status code check in Soneso/StellarSDK/Requests/RequestBuilder.php:
(so it never hits response handler)

```

 public function executeRequest(string $url, string $requestType, ?string $requestMethod = "GET") : Response {
        $response = null;
        try {
            $request = new Request($requestMethod, $url, RequestBuilder::HEADERS);
            $response = $this->httpClient->send($request);
        }
        catch (GuzzleException $e) {
            throw HorizonRequestException::fromOtherException($url, $requestMethod, $e, $response);
        }

        if (300 <= $response->getStatusCode()) {
                throw new HorizonRequestException('error in response: '.$response->getStatusCode());
        }

        $responseHandler = new ResponseHandler();
        try {
                return $responseHandler->handleResponse($response, $requestType, $this->httpClient);
        } catch (Exception $e) {
				throw HorizonRequestException::fromOtherException($url, $requestMethod, $e, $response);
        }
}

```

Now It throws HorizonRequestException since that is what is supposed to throw, according to the comments in the code.

Maybe there's a better way?

Thank you.

**Comment by christian-rogobete** [comment]
Hello @arduent,
thank you for adding this issue and improvement request.

As I understand you, the issue only occurs if you change the parameter from 'errors' to 'http_errors' in the SDK. We should update that but it has side effects. Meaning that we need to find and remove all of them if we do so.

I like your approach and if you would like to adjust the necessary places, it would be a great improvement for the SDK. I would be more than happy to review and merge your pull request. Pls. let me know.

Btw. you can check if the account is is valid before sending it to horizon. Just try to create a keypair from it. If you get InvalidArgumentException you know that it is invalid and do not have to send the request to horizon.

**Comment by arduent** [comment]
Thanks for the info about checking key validity. I will update the SDK code and put in a merge request.

**Comment by christian-rogobete** [comment]
Resolved in [release 1.5.3](https://github.com/Soneso/stellar-php-sdk/releases/tag/1.5.3).
